### PR TITLE
Stabilize `#![feature(const_if_match)]` and `#![feature(const_loop)]`

### LINKED ISSUE #49146: Tracking issue for RFC 2342, "Allow `if` and `match` in constants"
This is a tracking issue for the RFC "Allow `if` and `match` in constants" (rust-lang/rfcs#2342).

**Please redirect constification of specific functions or issues you want to report to fresh issues and label them appropriately with `F-const_if_match` so that this issues doesn't get flooded with ephemeral comments obscuring important developments.**

**Steps:**

- [x] Implement the RFC
- [ ] Adjust documentation ([see instructions on forge][doc-guide])
- [ ] Stabilization PR ([see instructions on forge][stabilization-guide])
- [ ]  `let` bindings in constants that use `&&` and `||` short circuiting operations. These are treated as `&` and `|` inside `const` and `static` items right now.

[stabilization-guide]: https://forge.rust-lang.org/stabilization-guide.html
[doc-guide]: https://forge.rust-lang.org/stabilization-guide.html#updating-documentation

**Unresolved questions:**

None

### LINKED ISSUE #66753: Can't implement a const `Option::unwrap` because of "destructors cannot be evaluated at compile-time" error.
It's not possible to implement `Option::unwrap` as a const fn like:
https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=8ad0c1cdd371ca0488fab9d872bd6ffe
```rust
#![feature(const_fn)]
#![feature(const_if_match)]
#![feature(const_panic)]

const fn unwrap<T>(opt:Option<T>)->T{
    match opt {
        Some(x)=>x,
        None=>panic!("Trying to unwrap a None"),
    }
}
```

Because the function triggers this error:
```
error[E0493]: destructors cannot be evaluated at compile-time
 --> src/lib.rs:5:20
  |
5 | const fn unwrap<T>(opt:Option<T>)->T{
  |                    ^^^ constant functions cannot evaluate destructors

error: aborting due to previous error
```

### PR DESCRIPTION
**Pull request by ecstatic-morse**
Quoting from the [stabilization report](https://github.com/rust-lang/rust/issues/49146#issuecomment-616301045):

> `if` and `match` expressions as well as the short-circuiting logic operators `&&` and `||` will become legal in all [const contexts](https://doc.rust-lang.org/reference/const_eval.html#const-context). A const context is any of the following:
>
> - The initializer of a `const`, `static`, `static mut` or enum discriminant.
> - The body of a `const fn`.
> - The value of a const generic (nightly only).
> - The length of an array type (`[u8; 3]`) or an array repeat expression (`[0u8; 3]`).
>
> Furthermore, the short-circuiting logic operators will no longer be lowered to their bitwise equivalents (`&` and `|` respectively) in `const` and `static` initializers (see #57175). As a result, `let` bindings can be used alongside short-circuiting logic in those initializers.

Resolves #49146.

Ideally, we would resolve :whale: #66753 before this lands on stable, so it might be worth pushing this back a release. Also, this means we should get the process started for #52000, otherwise people will have no recourse except recursion for iterative `const fn`.

r? @oli-obk

### CODE REVIEW

**Comment by ecstatic-morse** [comment]
While writing this, I found and fixed a bug in the HIR const-checker where it wasn't requiring [`allow_internal_unstable`](https://github.com/rust-lang/rust/pull/72437/files#diff-bc6ec17320b71b69c208fd08521e93edR9) for `rustc_const_stable` functions, just the feature gates. I will probably split this out as part of an upcoming refactoring of the `fn_queries` module.

**Comment by Manishearth** [comment]
ping @oli-obk

**Comment by oli-obk** [comment]
This is waiting for someone from @rust-lang/lang starting the FCP on https://github.com/rust-lang/rust/issues/52000 as we (wg-const-eval) would like to stabilize looping at the same time as branching so ppl don't have to resort to recursion.

**Comment by Manishearth** [comment]
Ah, gotcha. Thanks!

**Comment by KodrAus** [comment]
There’s now a [FCP](https://github.com/rust-lang/rust/issues/52000#issuecomment-641160825) on #52000

**Comment by jhpratt** [comment]
*proposed FCP. Still needs a few checkboxes.

**Comment by oli-obk** [comment]
Please adjust this PR to also stabilize `const_loop`. The FCP on the tracking issue has been started, imo we should get this merged immediately and revert if the FCP leads to blocking concerns before it finishes.

**Comment by mark-i-m** [comment]
FCP on https://github.com/rust-lang/rust/issues/52000 has completed!

**Comment by Manishearth** [comment]
@bors r=oli-obk

(this was included in the rollup)

**Comment by Dylan-DPC** [comment]
failed in [rollup CI](https://github.com/rust-lang/rust/pull/73821#issuecomment-650653860)

**Comment by CryZe** [comment]
This PR states that they are stabilized in 1.45, but that's long been in beta and the milestone for the PR is 1.46. That's a mistake right?

File: src/librustc_feature/accepted.rs [hunk 1]
```diff
@@ -261,6 +261,10 @@ declare_features! (
     (accepted, transparent_enums, "1.42.0", Some(60405), None),
     /// Allows using subslice patterns, `[a, .., b]` and `[a, xs @ .., b]`.
     (accepted, slice_patterns, "1.42.0", Some(62254), None),
+    /// Allows the use of `if` and `match` in constants.
+    (accepted, const_if_match, "1.45.0", Some(49146), None),
+    /// Allows the use of `loop` and `while` in constants.
+    (accepted, const_loop, "1.45.0", Some(52000), None),
```
Comment (CryZe) [inline src/librustc_feature/accepted.rs:267]: In particular here.

**Comment by Dylan-DPC** [comment]
Yes.
The milestone wasn't updated (I forgot to do it before the previous release) and since this PR was started before the previous release, it should have been updated. We can update it in a follow up pr

**Comment by ehuss** [comment]
@ecstatic-morse or @oli-obk would you be able to send a PR to update the reference [const eval](https://github.com/rust-lang/reference/blob/master/src/const_eval.md) chapter with these changes (and anything else that might have been missed since it was last updated)?

**Comment by oli-obk** [comment]
Oh yea. Good idea, thanks! Doing this right away.

**Comment by Dylan-DPC** [comment]
I was looking to offer it to some newcomer :P

**Comment by oli-obk** [comment]
oops, too late: https://github.com/rust-lang/reference/pull/842

**Pull request merged**